### ISSUE TITLE
improve our overlay

### ISSUE THREAD

**Issue by alexander-akait** [webpack/webpack-dev-server#3689]
- [ ] This is a **bug**
- [x] This is a **modification** request

### Code

improve our default overlay:
1. Allow to catch `runtime` errors
2. Improve errors/warnings reporting (ideally we should have )
3. Improve styles
4. More?

CRA has `react-error-overlay`, I think we can union with them and create the one great overlay

/cc @raix  

### Expected Behavior

Better DX

### Actual Behavior

Our overlay is very simple and in some places have bad DX

**Comment by Bosch-Eli-Black** [comment]
Not sure if it's possible, but it would be awesome if compilation errors were shown after refreshing the page! 🙂  Currently if the page is refreshed when there's a compilation error, we just get a whitescreen and no overlay is shown.

**Comment by SuneelFreimuth** [comment]
I'd be happy to try implementing these improvements.

**Comment by alexander-akait** [comment]
> Not sure if it's possible, but it would be awesome if compilation errors were shown after refreshing the page! slightly_smiling_face Currently if the page is refreshed when there's a compilation error, we just get a whitescreen and no overlay is shown.

Can you provide reproducible steps, I think it is bug

**Comment by ylemkimon** [comment]
`react-error-overlay` seems to be easily integrable, adding it to the client works out of the box. Two notes:
- it doesn't support build warnings, i.e., build errors are not dismissable
- it also displays runtime errors, which is not the existing overlay did. If the dev server decides to support this, source code fetch logic should be also updated

**Comment by alexander-akait** [comment]
@ylemkimon yep, feel free to start it, for runtime errors we need a new options for overlay, i.e. `overlay.runtimeErrors`

**Comment by sibelius** [comment]
how can I customize overlay behavior?

I've tried this

````jsx
devServer: {
   overlay: {
      warnings: false,
      errors: true
    }
}
```

but it did not worked

**Comment by alexander-akait** [comment]
Should work, what is the problem?

**Comment by ylemkimon** [comment]
@sibelius If you're using dev server v4, you should put `overlay` in `client` (https://github.com/webpack/webpack-dev-server/blob/master/migration-v4.md):
```js
devServer: {
   client: {
      overlay: {
         warnings: false,
         errors: true
       }
   }
}
```

**Comment by malcolm-kee** [comment]
I think all the 3 listed can be implemented independently.

I would like to share my current thought on number 3 (improve styles). 

Based on what I've seen in `react-error-overlay`, it uses React to render the overlay, I think we should avoid bringing in React just for overlay.

Instead, we can stick with current approach (manually constructing the HTML using plain JS), and just port-in the style.

**Comment by alexander-akait** [comment]
@malcolm-kee What we need to implement more, and I will update our check list or maybe we need a small dicussion about other features from CRA (or other good overlays)

**Comment by malcolm-kee** [comment]
CRA error stack was pretty nice as it showed code snippet, but I not sure if that's possible given that we don't have control over `devtool` options. 

Other than that I'm thinking of making runtime overlay less intrusive, like showing at bottom right, since it's usually doesn't break application code like compilation error does.

**Comment by alexander-akait** [comment]
@malcolm-kee 

> Other than that I'm thinking of making runtime overlay less intrusive, like showing at bottom right, since it usually doesn't break application code like compilation error does.

Sounds good, even more, I think we can add an option for this, even for compilation errors/warnings

> CRA error stack was pretty nice as it showed code snippet, but I not sure if that's possible given that we don't have control over devtool options.

Can you show me, because we can use generate source maps (if they were enabled)

**Comment by malcolm-kee** [comment]
> Can you show me, because we can use generate source maps (if they were enabled)

Found the following gif from [here](https://blog.jakoblind.no/error-overlay-with-react-hot-loader/).

![cra-runtime-error](https://user-images.githubusercontent.com/24528512/236681895-04dbb091-3a0e-4673-8c15-6775ae64a21b.gif)

**Comment by alexander-akait** [comment]
hm, I see, I think we can improve it on webpack side, I will look deeply on this before preparing the new release and will say how we can achive it

**Comment by malcolm-kee** [comment]
Another inspiration from React Native:

![React Native Error](https://user-images.githubusercontent.com/17029084/56429455-317e0d00-62c3-11e9-8a32-e7f95c34764f.png)

We can probably support shortcut key so it's easy to show/hide the error.

**Comment by malcolm-kee** [comment]
Adding some design: https://www.figma.com/file/xyeFjFzImMya1MdDDn7vii/Webpack-Error-Overlay?type=design&node-id=0%3A1&t=SgABTKrN4qJU6k15-1

Feedbacks are welcomed.

## Compilation error

![compilation-error](https://github.com/webpack/webpack-dev-server/assets/24528512/dd4ea96b-25e6-4279-b81a-f70c5b86c6ec)

## Runtime error notification

This notification button will be positioned fixed at bottom right

![runtime-notification](https://github.com/webpack/webpack-dev-server/assets/24528512/d31a97ce-2f3a-4b0d-9042-23fae25bdb4e)

## Runtime error (when clicked)

![runtime-error](https://github.com/webpack/webpack-dev-server/assets/24528512/643b2cc5-b484-47e8-9c9e-b71cd4daa6e9)

**Comment by levrik** [comment]
@malcolm-kee These designs look awesome. Sad that it's unlikely they will ever get implemented.

**Comment by alexander-akait** [comment]
@levrik I am fine with such improvements and the new design, if you want to send PR - welcome

**Comment by snitin315** [comment]
I can try implementing this design post major version release.

**Comment by levrik** [comment]
@alexander-akait I would love to but doubt I'll find the time for it.
@snitin315 This would be awesome!
